### PR TITLE
Feature/permanent dns

### PR DESCRIPTION
**Pull request by pappz**
## Describe your changes

Because on Android system hard to update DNS settings the system will set our DNS service as permanent DNS server on mobile system.
This PR modify the original logic. Now the DNS service can functional in two different way
1. the DNS service will start listen when get update from mgm server
2. the DNS service start listening immediately when it instantiated

In the second case the upstream DNS server will be the DNS servers from the host machine.
The relevant Android code: https://github.com/netbirdio/react-native-app/pull/24

### Android
The Android code track the network changes and update the DNS configuration when it has been changed on the host machine. For example when switch from 4G to WIFI.

### Test
Because this change refactor the whole logic necessary to test all combination of the supported systems
- Mac, Win, Linux, Android
- On Linux with WG bind interface and with kernel module support
- On Linux with systemd, resolve.conf

### Checklist
- [ ] Is it a bug fix
- [ ] Is a typo/documentation fix
- [x] Is a feature enhancement
- [ ] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: client/internal/dns/server_test.go [hunk 1]
```diff
@@ -478,7 +522,8 @@ func TestDNSServerStartStop(t *testing.T) {
 func TestDNSServerUpstreamDeactivateCallback(t *testing.T) {
 	hostManager := &mockHostConfigurator{}
 	server := DefaultServer{
-		dnsMux: dns.DefaultServeMux,
+		//dnsMux: dns.DefaultServeMux,
```
Comment (pascal-fischer) [inline client/internal/dns/server_test.go]: some comment left

Comment (pappz) [reply client/internal/dns/server_test.go]: Hmm, interesting. I can not remember well why I did it. I will double check this test.

**Review by pascal-fischer — APPROVED** [review]

File: client/android/dns_list.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+// Add new DNS address to the collection
+func (array *DNSList) Add(s string) {
+	array.items = append(array.items, s)
+}
+
+// Get return an element of the collection
+func (array *DNSList) Get(i int) string {
+	return array.items[i]
```
Comment (gigovich) [inline client/android/dns_list.go]: We can get panic here, better to check the range of i.

**Review by gigovich — APPROVED** [review]
LGTM

File: client/android/dns_list.go (hunk 2 above)
Comment (pappz) [reply client/android/dns_list.go]: Thank you! I pushed the fix and I also updated the Android code

**Review by mlsmaycon — APPROVED** [review]

**Review by mlsmaycon — APPROVED** [review]

**Pull request merged**